pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -138,6 +148,73 @@ type portToSelectorAllow map[uint16]cachedSelectorREEntry
 // policy.L7DataMap but the DNS rules are compiled into a single regexp
 type cachedSelectorREEntry map[policy.CachedSelector]*regexp.Regexp
 
+// structure for restored rules that can be used while Cilium agent is restoring endpoints
+type perEPRestored map[uint64]restore.DNSRules
+
+// CheckAllowed checks endpointID, destPort, destID, and name against the rules
+// added to the proxy, and only returns true if this all match something that
+// was added (via SetAllowed) previously.
+func (p *DNSProxy) checkRestored(endpointID uint64, destPort uint16, destIP string, name string) bool {
+	ruleIPs, exists := p.restored[endpointID][destPort]
+	if !exists {
+		return false
+	}
+
+	for re, ips := range ruleIPs {
+		if _, exists := ips[destIP]; exists && re.MatchString(name) {
+			return true
+		}
+	}
+	return false
+}
+
+// GetRules creates a fresh copy of EP's DNS rules to be stored
+// for later restoration.
+func (p *DNSProxy) GetRules(endpointID uint16) restore.DNSRules {
+	p.Lock()
+	defer p.Unlock()
+
+	restored := make(restore.DNSRules)
+	for port, entries := range p.allowed[uint64(endpointID)] {
+		ruleIPs := make(restore.RuleIPs)
+		for cs, regex := range entries {
+			var IPs map[string]struct{}
+			if !cs.IsWildcard() {
+				IPs = make(map[string]struct{})
+				for _, nid := range cs.GetSelections() {
+					for _, ip := range p.LookupIPsBySecID(nid) {
+						IPs[ip] = struct{}{}
+					}
+				}
+			}
+			ruleIPs[restore.RuleRegex{Regexp: regex}] = IPs
+		}
+		restored[port] = ruleIPs
+	}
```

Comment:
Yeah so the interesting question to me is how this differs from the incoming `L7DataMap` form. Practically this would be the full accumulated set of DNS rules acquired to date in the proxy which would be serialized out. Then we're using the selector cache to dump the current evaluations of IPs/identities selected.

For policies where it's just DNS allow-all this should be relatively straightforward (`cs.IsWildcard()` => true) but if you ended up with large numbers of DNS rules it could take some time, any/every time the DNS policy is updated. Would be a good candidate for a metric so we can find out how performance turns out and whether we need to further optimize.

Reply:
restore.RuleIPs here is already arranged so that allowed IPs are stored for each regex, while in 'p.allowed' regexes are behind selectors. Here we need to expand the selected IPs for each selector, so the nesting is unavoidable.

In the recommended configuration there would be only one selector (for the kube dns servers) which should only have a small set of IPs.  Wildcard selector is already optimized so that no IPs need to be listed.

Maybe some limits could be applied to guard against pathological policies here.

Reply:
@jrajahalme ^ Please do correct anything I got wrong above :)